### ISSUE TITLE
About the BinOp&Training

### ISSUE THREAD

**Issue by Paul0629** [jiecaoyu/XNOR-Net-PyTorch#29]
The BinOp helps to binarize the weights and store them in self.saved_params. But the weights in model remains the same.  'output = model(data)' in the training still use the floating point weights.  So where do we perform the bitcount or binary weights? @jiecaoyu  Really appreciate ur help!     
"       optimizer.zero_grad()
        bin_op.binarization()
        output = model(data)
        loss = criterion(output, target)
        loss.backward()
        bin_op.restore()
        bin_op.updateBinaryGradWeight()
        optimizer.step()"

**Comment by akhauriyash** [comment]
The bitcount is only needed for high performance kernels. 
This implements the algorithm without the speed up offered by the binary GEMM kernels. 
It is a demonstration of concept of sorts. 
You can check out https://github.com/MatthieuCourbariaux/BinaryNet for the Run Time implementation. But this repository is better -- it has convolution too.

**Comment by jiecaoyu** [comment]
@akhauriyash Thanks so much for helping answer the question!

@Paul0629 Hi, as @akhauriyash said, this repo provides a solution for training. For real inference, it needs to be combined with external computation kernels.
